File: leetcode/src/main/golang/valid_parentheses.go
func isValid(s string) bool {
    if len(s) % 2 == 1 {
        return false
    }
    var stack []rune
    for _, c := range s {
        if len(stack) != 0 && stack[len(stack) - 1] == c {
            stack = stack[0:len(stack) - 1]
            continue
        }
        if c == '(' {
            stack = append(stack, ')')
        } else if c == '[' {
            stack = append(stack, ']')
        } else if c == '{' {
            stack = append(stack, '}')
        } else {
            return false
        }
    }
    return len(stack) == 0
}

File: leetcode/src/main/ruby/valid_parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    return false if s.length & 1 == 1
    stack = []
    s.each_char { |c|
        if stack.length != 0 && stack[-1] == c
            stack = stack[0...-1]
            next
        end
        if c == '('
            stack << ')'
        elsif c == '['
            stack << ']'
        elsif c == '{'
            stack << '}'
        else
            return false
        end
    }
    return stack.length == 0
end